Logger app using `logging`

Aijuh opened the pull request:
This pull request is about #181

I set logger in logger.py as root logger, and logger in qiwis.py as __name__(__main__), so log from qiwis.py propagate to logger.py. Made loggingHandler to deal with log message and print to loggerFrame window. log message from qiwis.py printed in logger Frame.
Moreover I additionally add new Frame for selecting logging level (ex. DEBUG, ERROR .. ). If it is not needed, I will delete that part of code and pull request again.
Thank you.

Aijuh committed:
enable logger

In qiwis.py fix logger name as "parent"
In examples/logger.py fix followings.

1. import logging
2. set logger name as "parent" and setLevel as INFO to take msg from qiwis (qiwis using log.info)
3. (line 17-28) define signal and Qthandler.
4. (line 105-110) define formatter and  add handler to logger
5. (line 116) add pyqtSlot(str)

Aijuh committed:
using root logger and add button for selecting logger level

1. In qiwis.py and logger.py, changed logger as root logger by using getLogger()
In logger.py followings are changed
2. change Qthandler and Signaller class to one LoggingHandler class (line 16-24)
3. new button for selecting logger level is added by line 40, 45
4. implement frame for select logger level (line 82-120)
5. In loggerApp, make code for connecting logger level button and logger level Frame. (line 141-142, 180-183)

Aijuh committed:
.

.

Aijuh committed:
change logger name in qiwis, add comment

In qiwis, change logger name to __name__

In logger.py add comment for added code.

Aijuh committed:
Update logger.py

kangz12345 commented on examples/logger.py:
```diff
@@ -3,13 +3,33 @@
 
 from qiwis import BaseApp
 
+logger = logging.getLogger()
```
Actually, we just want to add a handler to the root logger.
We don't have to define it as a global variable here.

kangz12345 commented on examples/logger.py:
```diff
@@ -3,13 +3,33 @@
 
 from qiwis import BaseApp
 
+logger = logging.getLogger()
+logger.setLevel(logging.INFO)
```
We should not change the root logger level. We only handle the log records. Changing the root logger might results in unexpected behaviors when using with other modules.

kangz12345 replied:
Therefore, please remove these two lines!

kangz12345 commented on examples/logger.py:
```diff
@@ -3,13 +3,33 @@
+logger.setLevel(logging.INFO)
+
+class LoggingHandler(QObject, logging.Handler):
+    """
+    Handler for logger, handle input log by emit
+    """
```
We don't break lines for a single line docstring, for example:
`"""Handler for logger, handle input log by emit."""`

In addition, please finish your sentence with a period(.).

kangz12345 replied:
However, since we write a docstring for public signals, you should write it for `signal`.
For example:
```python
class LoggingHandler(...):
    """Handler for logger ...

    Signals:
        signal(log): A formatted log message is emitted.
    """
```

You can check other example docstrings in qiwis.py as well (e.g., `Qiwis`).

kangz12345 commented on examples/logger.py:
```diff
@@ -3,13 +3,33 @@
+    """
+    Handler for logger, handle input log by emit
+    """
+    signal = pyqtSignal(str)
```
Please make a single empty line after the class docstring.

kangz12345 commented on examples/logger.py:
```diff
@@ -3,13 +3,33 @@
+    Handler for logger, handle input log by emit
+    """
+    signal = pyqtSignal(str)
+    def __init__(self, slotfunc, *args, **kwargs):
```
Please make a single empty line between class-level definitions (after `signal = pyqtSignal(str)`).

kangz12345 commented on examples/logger.py:
```diff
@@ -3,13 +3,33 @@
+logger = logging.getLogger()
+logger.setLevel(logging.INFO)
+
+class LoggingHandler(QObject, logging.Handler):
```
Of course you might want to make the code simpler by merging the `Signaller` in the example with the handler.
However, multiple inheritance could bring us subtle problems such as confusing method name resolution, and `super()` calls.
I would recommend to separate the `QObject` class and `logging.Handler` class, just like the example.

BECATRUE replied:
I also agree with @kangz12345.

BECATRUE replied:
@Aijuh Please modify this part!

kangz12345 commented on examples/logger.py:
```diff
@@ -3,13 +3,33 @@
+    """
+    signal = pyqtSignal(str)
+    def __init__(self, slotfunc, *args, **kwargs):
+        super().__init__(*args, **kwargs)
```
From this code, the reader is confused whether it calls the `__init__` of `QObject` or `logging.Handler`. If you inherit only one class, then the problem is resolved.

kangz12345 commented on examples/logger.py:
```diff
@@ -3,13 +3,33 @@
+        self.signal.connect(slotfunc)
+
+    def emit(self, record):
+        """ 
+        emit input signal to connected function
+        """
```
1. Do not break the line after the first `"""`.
2. Use a capital letter for the first character.
3. Use the 3rd person singular verbs (emits, not emit) for function/method docstrings. It is easy to think the subject of the sentence is "it" and you omit it.
4. Finish sentences with a period.
5. Write the docstring for arguments.

Following the [style guide](https://google.github.io/styleguide/pyguide.html), the docstring should be:
```python
        """Emits input signal to connected function.
        
        Args:
            record: The emitted log record.
        """
```

However, since this is _overriding_ the method of `logging.Handler`, we can write as (see also: #44):
```python
        """Overridden."""
```

kangz12345 commented on examples/logger.py:
```diff
@@ -64,7 +86,48 @@ def buttonCancelClicked(self):
         self.close()
 
+class SetLevelFrame(QWidget):
+    """
+    Extended.
+    """
+    def __init__(self, parent: Optional[QObject] = None):
+
```
You should write the `"""Extended."""` docstring in the `__init__()`.

kangz12345 commented on examples/logger.py:
```diff
@@ -85,11 +148,21 @@ def __init__(self, name: str, parent: Optional[QObject] = None):
+        fs ="%(name)s %(message)s"
+        formatter = logging.Formatter(fs)
+        self.handler.setFormatter(formatter)
+        logger.addHandler(self.handler)
```
Note that you only need the root logger object here.
Therefore, you can just write:
```python
logging.getLogger().addHandler(self.handler)
```

kangz12345 commented on qiwis.py:
```diff
@@ -37,7 +37,6 @@
 
 logger = logging.getLogger(__name__)
 
-
```
Please discard this change since it follows the rule (2 empty lines between global definitions).

kangz12345 replied:
Please discard this change.

kangz12345 commented on examples/logger.py:
```diff
@@ -115,6 +188,11 @@ def receivedSlot(self, channelName: str, content: Any):
+            
+    @pyqtSlot()
+    def selectLevel(self):
+        """"shows a level select frame"""
```
1. Start with a capital letter.
2. Finish with a period.

kangz12345 reviewed (changes requested):
Great job implementing the log level selection frame as well, but it is not a part of this PR and seems not that simple feature (there are some more design issues to decide). Please remove that feature in this PR.

In addition, please use [type annotations](https://peps.python.org/pep-0484/) in function or method definitions.

Aijuh commented:
Thank you. 
I will follow your instruction and fix the code.

Aijuh committed:
feedback included

1. Seperate Signaller and Handler.

2. Follows comment rule.

3.  logger is not global variable in logger.py

4. delete logger level selection app which is not needed

Aijuh committed:
add comment for new class "Signaller"

.

Aijuh committed:
Update logger.py

Aijuh committed:
Update logger.py

Aijuh committed:
Update logger.py

Aijuh commented:
I changed my code and update it.
Should I make another pull request or just use this pull request?

kangz12345 commented:
You can stay in this PR. The reviewers review and you reply to that and apply the reviews, and we repeat this process. At the end, when the reviewers approve this PR, then you can merge it.

kangz12345 commented on examples/logger.py:
```diff
@@ -3,13 +3,38 @@
 from qiwis import BaseApp
 
+class Signaller(QObject):
+    """Signal for LoggingHandler"""
```
Period at the end!

BECATRUE commented on examples/logger.py:
```diff
@@ -3,13 +3,38 @@
+class LoggingHandler(QObject, logging.Handler):
+    """Handler for logger.
+
+    Sends log message to connected function using emit
```
It's just an option, but please add 'a' or 'the' properly.

BECATRUE replied:
Also, please add the period at the end.

Aijuh replied:
I missed it. Thank you.

BECATRUE commented on examples/logger.py:
```diff
@@ -29,7 +54,6 @@ def __init__(self, parent: Optional[QObject] = None):
         layout.addWidget(self.logEdit)
         layout.addWidget(self.clearButton)
 
-
```
Please add _two_ empty lines between outermost classes and functions!
For example, refer to other [files](https://github.com/snu-quiqcl/qiwis/blob/develop/qiwis.py).

BECATRUE commented on examples/logger.py:
```diff
@@ -85,11 +108,19 @@ def __init__(self, name: str, parent: Optional[QObject] = None):
         self.confirmFrame.confirmed.connect(self.clearLog)
+        # define handler
+        self.handler = LoggingHandler(self.addLog)
+        # arbitrary format
```
If you plan to update this format, please use _TODO_ docstring referred to the [styleguide](https://google.github.io/styleguide/pyguide.html#312-todo-comments).

kangz12345 replied:
Please make this a TODO comment, for example `# TODO(aijuh): Change the log format when it is determined.`

BECATRUE commented on examples/logger.py:
```diff
@@ -3,13 +3,38 @@
+    signal = pyqtSignal(str)
+
+    def __init__(self, slotfunc, *args, **kwargs):
+        super().__init__(*args, **kwargs)
```
First of all, please add the constructor docstring about arguments.
Also are both `args` and `kwargs` necessary?

Aijuh replied:
No, I just copy and paste example code and didn't check it. Thank you.

BECATRUE commented on examples/logger.py:
```diff
@@ -3,13 +3,38 @@
+    Sends log message to connected function using emit
+
+    Signal:
+          signal: Helps to send message to destination.
```
Where is the signal used? I wonder what you intended.

Aijuh replied:
Good review. I should delete it.

BECATRUE commented on examples/logger.py:
```diff
@@ -85,11 +108,19 @@ def __init__(self, name: str, parent: Optional[QObject] = None):
+        fs ="%(name)s %(message)s"
+        formatter = logging.Formatter(fs)
+        self.handler.setFormatter(formatter)
+        logging.getLogger().addHandler(self.handler)
```
Is the logger in `qiwis.py` being called in well?

It seems not to print the logs generated in `qiwis`.
Also I think an appropriate name is necessary in `getLogger()`.

Aijuh replied:
In order to print log message, the logger level should be set as logging.INFO in logger.py, because logger.info is used in qiwis.py to propagate log message.

From @kangz12345 I told to test root logger, so I just used root logger. But in order to manage logger level like INFO, DEBUG, it seems that name needed.

 I will set the name as 'parent' temporarily, and change logger level to show you message is well printed in next update.

BECATRUE replied:
Oh, you're right! I didn't know `logging.getLogger()` would return the root logger.
Thanks for your detailed description😃

BECATRUE reviewed (changes requested):
Good start! I left several reviews.

I couldn't check if the logging is working well because it doesn't show the logs created in `qiwis`.
Let me know if I did anything wrong.

Aijuh committed:
Include feedback from BECATRUE

1. Maintain 2 blank lines between classes.

2. In Logging handler's __init__, delete unused arguments and update comment.

3. In loggerApp in logger.py, logger name is set as "parent". Also, logger name in qiwis.py is changed to set it as child logger of logger in logger.py.

Aijuh committed:
Update logger.py

Aijuh committed:
Update logger.py

Aijuh commented:
Thank you for the detailed review. I fixed the code.

Aijuh committed:
Update logger.py

BECATRUE reviewed (changes requested):
Please check the unresolved reviews written by Jiyong.

Aijuh commented:
I understand that we should not change root logger's level to prevent error. 
So I set logger name as 'parent'(name can be changed depend on notation rule) and upload  the code.
`
        logger = logging.getLogger("parent")
        logger.setLevel(logging.INFO)
        logger.addHandler(self.handler)
`
logger level is temporarily set as logging.INFO to print logger.info in qiwis.

Aijuh committed:
Update logger.py

Aijuh committed:
Update logger.py

Aijuh committed:
Update logger.py

BECATRUE commented:
Can you explain what you planned about a logger name and logging level?
I think we should decide them first.

In my opinion, I think we can just define a logger in each module such as `qiwis.py` and each apps as follows.
```python
logger = logging.getLogger(__name__)
```
It is the recommended method in the [docs](https://docs.python.org/3/howto/logging.html#advanced-logging-tutorial).

In terms of a logging level, I think each module is responsible for its logger and in `logger.py` it should just show the passed logs.
If we want to set the minimum logging level in the logger GUI, we can use `self.handler.setLevel()` referred to the [docs](https://docs.python.org/ko/3/library/logging.html#logging.Handler.setLevel).

I wonder what @kangz12345 intended and how @Aijuh are going to implement.

Aijuh commented:
I am planning to set logger in logger.py as "parent" and set other module's logger using `logging.getLogger("parent").getchild(__name__)` (might be changed).
 If I set logger name like this, logger from child propagate to parent. It works well between logger.py and qiwis.py.
 Of course I should test for other modules later.

For setting logger level, what I understand about @BECATRUE 's suggestion is each module is responsible for its logger and logger level. In this case what I am planning is as follows. 
1. Set logger level of parent logger in logger.py as DEBUG. Then, any logger message from other module will be printed.
2. Set logger level of child logger separately in that module. If we set logger level as ERROR, and input logger as logger.info, the message didn't propagate to parent logger in logger.py.

In case of changing handler's level using `self.handler.setLevel()`, it can block INFO level logs from DEBUG level logger. But it cannot allow ERROR level logger to get DEBUG level logs. 
In other words, it cannot lower logger's level. 
So I think that we may change logger's level directly or fix logger's level as DEBUG, and change handler's level.

I will wait exact name convention of logger from @kangz12345 and @BECATRUE .

BECATRUE commented:
Thanks for your detailed explanation👍

First, I strongly agree with your idea about setting the logger levels.
Also setting all loggers as subloggers of a specific logger, i.e. `parent` logger is the best solution in terms of management and danger of setting the level of the root looger.

However, I think it is a problem that `logger.py`, not `qiwis.py` has the authority to set up the `parent` logger.
Also it is a big restriction that all users using this package should know that all loggers must be subloggers of the `parent` logger.
So I suggests to move the authority to `qiwis.py` as follows.
```python
# qiwis.py
qiwisLogger = logging.getLogger("parent")
logger = qiwisLogger.getChild(__name__)  # Same as the current logger

# logger.py
import qiwis
...
logger = qiwis.qiwisLogger

# Other apps
import qiwis
...
logger = qiwis.qiwisLogger.getChild(__name__)
```
The `logger.py` is just an app, not a part of `qiwis` package, so we should remove any dependence on `qiwis.py`'s position.

Finally, about the name of the parent logger, I think 'qiwis' is more appropriate because `parent` looks a liitle ambiguous.

After listening to Jiyong's opinion, let's make a final decision.

Aijuh committed:
discard unnecessary inheritance

kangz12345 commented:
Thank you for the discussion.
I also talked with @Aijuh offline, and we decided to do one of the following:
### 1. Use the root logger and modify its level.
This might cause some unexpected behavior when any other module uses the root logger as well.
However, such a module also takes a risk that the root logger can be used by any other module.
Therefore, we just need to document that this modifies the root logger level.

### 2. Construct a logger hierarchy and make the apps follow the naming convention.
This does not modify the root logger, so it can avoid the problem of the option 1.
However, it cannot catch the logs emitted by the loggers which do not follow the naming convention (consider the loggers of any third-party modules which do not care of `qiwis`).
Moreover, the way to set the logger to be a child logger of `qiwis` logger is quite inconvenient.

In my opinion, unless there is a convenient way to set the logger hierarchy and catch the third-party apps' logs and there is no more problem with modifying the root logger, the option 1 looks better.

BECATRUE commented:
Thanks for suggesting options!

I don't think setting the child logger in option 2 is convenient because we can make a helper function in `qiwis.py`.
But the issue looks critical that we cannot track the logs generated by several third-party modules.

Thus I also agree with option 1.

Aijuh commented:
Getting discussion with @kangz12345 at yesterday lab meeting, what I understood is as follows.
We should consider the case where other App importing qiwis and also importing other module (for ex temp), and that temp module also using logging system. We want to manage this logger message from temp too.
In that case, using naming convention for parent and child is useless because temp will not know that convention.

So The solution is like this.

1. Using root logger in logger.py
2. Not using root logger. 

In 1. we should change root logger's level as DEBUG and manage logger.handler to block message. This solution is quite simple but dealing with root logger may cause problem. There is no serious problem in codes, but if other module also using root logger and we change root logger's level, it would cause problem. Also if we set root logger's level as DEBUG, all logger's default level is set as DEBUG.  

In 2. what I am planning is using naming convention of parent and child relationship in qiwis Apps. And at specific app who wants to read imported module's log we can manually read message and deliver to app's logger. logger.py's logger is the highest parent, and qiwis.py will be its child. And all apps logger is child of qiwis.py's logger. 

But I don't know detailed rules for importing qiwis.py in other module. If I got something wrong please correct it. Thank you.

Aijuh commented:
I understand there are two options.

1. Using root logger in logger.py
2. Not using root logger.
3. 
In 1. we should change root logger's level as DEBUG and manage logger.handler to block message. This solution is quite simple but dealing with root logger may cause problem. There is no serious problem in codes, but if other module also using root logger and we change root logger's level, it would cause problem. If we set root logger's level as DEBUG, all logger's default level is set as DEBUG, which is not a severe problem.

In 2. what I am planning is using naming convention of parent and child relationship in qiwis Apps. And at specific app who wants to read imported module's log we can manually read message and deliver to app's logger. logger.py's logger is the highest parent, and qiwis.py will be its child. And all apps logger is child of qiwis.py's logger.

As @kangz12345 told me that it may be okay to use root level logger and @BECATRUE said that it would be much simpler way, I fixed the code in that way. The root logger's level is now DEBUG. Please inform me if I missed something. Thank you.

Aijuh committed:
using root logger

1. Change logger in logger.py as root logger.

2. Change logger in qiwis.py as mere __name__(__main__)

Now logger in logger.py get all log from logging lib.

BECATRUE commented:
Good! And please apply the above reviews about code format and `Signaller` first.
Also, our final PR must meet the GitHub actions i.e. Pylint and unittest.
To check the failed cause, please click the 'Details' button in the below GitHub actions.

Aijuh committed:
Fix code to follow format for comment

Aijuh committed:
blank space error

Aijuh committed:
test for repeated error

Aijuh committed:
test for pyling error 2

Aijuh committed:
fix pylint error

Aijuh committed:
Update logger.py

Aijuh committed:
Update logger.py

Aijuh committed:
fixed pylint error

Aijuh committed:
Update logger.py

Aijuh commented:
Sorry for the redundant code change. Dealing with pylint.error, I misunderstood location of blank space.
I changed commitment to follow code format and add comment about signaller.

kangz12345 commented:
1. You can install `pylint` and test it with your own machine (you don't have to push and wait for the GitHub action to finish every time).
2. If you are done and want more feedback, you can re-request the review. If you don't have permission to request a review, please let me know.
3. Is it still not clear about the blank space rules, or now you understand it?

Aijuh commented:
> * You can install `pylint` and test it with your own machine (you don't have to push and wait for the GitHub action to finish every time).

I didn't know that, I will do in that way.


> * If you are done and want more feedback, you can re-request the review. If you don't have permission to request a review, please let me know.

I re-requested to reviewers.

> * Is it still not clear about the blank space rules, or now you understand it?

I understand it. I delete unnecessary blank space.

kangz12345 commented on examples/logger.py:
```diff
@@ -3,20 +3,51 @@
 import time
+import logging
 from typing import Any, Optional, Tuple
-
 from PyQt5.QtCore import QObject, pyqtSlot, pyqtSignal
 from PyQt5.QtWidgets import QWidget, QVBoxLayout, QPushButton, QTextEdit, QLabel, QDialogButtonBox
-
```
You should keep these blank lines.
Please read [PEP8](https://peps.python.org/pep-0008/#imports).
Also, I improved the developer guide #2.

kangz12345 commented on examples/logger.py:
```diff
@@ -3,20 +3,51 @@
 
+
+class Signaller(QObject):
+    """Signal for LoggingHandler."""
```
We put a single blank line after the class docstring.
I wrote this in #2.

Therefore:
```python
class Signaller(QObject):
    """Signal ..."""

    signal = pyqtSignal(str)


class LoggingHandler(...):
...
```

kangz12345 commented on examples/logger.py:
```diff
@@ -3,20 +3,51 @@
+
+    Sends a log message to connected function using emit.
+
+    Signals:
+        signal(log): A formatted log message is emitted.    
```
Please move this to the docstring of `Signaller`.

kangz12345 commented on examples/logger.py:
```diff
@@ -3,20 +3,51 @@
+    """
+
+    def __init__(self, slotfunc):
+        """Connect an input function to the signal.
```
You should start this docstring with "Extended" as described in #44.
For example:
```python
"""Extended.

Connects an input function to the signal.

Args:
    slotfunc: A slot function connected to the signaller's signal.
"""
```

kangz12345 replied:
In addition, please notice that "Connects" and "A slot ...".

kangz12345 replied:
You should start a sentence with a capital letter.
Please notice the argument docstring of `slotfunc`.
I changed some expression.

FYI: We call the functions bounded by a class instances _methods_.

kangz12345 replied:
Please read this @Aijuh

Aijuh replied:
I understand as adding argument docstring for `slotfunc`.

Since slotfunc for loggingHandler is addLog method in LoggerApp, I add that part.

kangz12345 commented on examples/logger.py:
```diff
@@ -3,20 +3,51 @@
+        self.signaller.signal.connect(slotfunc)
+
+    def emit(self, record: logging.LogRecord):
+        """ Emits input signal to connected function."""
```
Please remove the heading whitespace (here -> """_Emits).

kangz12345 commented on examples/logger.py:
```diff
@@ -85,11 +117,21 @@ def __init__(self, name: str, parent: Optional[QObject] = None):
         self.loggerFrame.clearButton.clicked.connect(self.checkToClear)
         self.confirmFrame = ConfirmClearingFrame()
         self.confirmFrame.confirmed.connect(self.clearLog)
+        # define handler
```
You don't have to write obvious comments (the code immediately tells you that a handler is defined).

kangz12345 commented on examples/logger.py:
```diff
@@ -85,11 +117,21 @@ def __init__(self, name: str, parent: Optional[QObject] = None):
+        self.handler.setFormatter(formatter)
+        logger = logging.getLogger()
+        logger.addHandler(self.handler)
+        logger.setLevel(logging.DEBUG)
```
You don't have to set the log level here. Let's set this later when you implement the level selector GUI.
It is okay that this cannot print out the info logs now.

Aijuh committed:
Reflect feedback from @kangz12345

Follow blank line and space rules.

logger level in logger.py is changed to default WARNING. If you want to print log, change it to DEBUG or INFO.

Aijuh commented:
I will read reference ([PEP8](https://peps.python.org/pep-0008/#imports), #44, #2) to follow code format. Nice catch for what I missed. 

I changed for what I missed and fix the codes.

kangz12345 commented:
Please check the unresolved comments.

Aijuh committed:
update unsolved issue

kangz12345 commented on examples/logger.py:
```diff
@@ -3,20 +3,57 @@
+        self.signaller.signal.connect(slotfunc)
+
+    def emit(self, record: logging.LogRecord):
+        """Emits input signal to connected function."""
```
Since this _overrides_ the method `logging.Handler.emit()`, you should start it with `"""Overridden."""`.
For example:
```python
"""Overridden.

Emits ...
"""
```

kangz12345 commented on examples/logger.py:
```diff
@@ -3,20 +3,57 @@
+        Connects the slotfunc to the signal.
+
+        Args:
+            slotfunc: method function connected to signal.
```
Please update this docstring as well.

Aijuh committed:
Update logger.py

Aijuh commented:
Add overridden message because loggingHandler in logger.py override logging.Handler.emit

Delete duplicated comments

Aijuh committed:
Update logger.py

kangz12345 commented on examples/logger.py:
```diff
@@ -3,20 +3,60 @@
+        Connects the slotfunc to the signal.
+
+        Args:
+            slotfunc: A method addLog in class LoggerApp.
```
Nice that you write it in a correct format, but the comment content is not appropriate.
The docstrings for public API (public classes, methods, functions, etc.) should be informative for users.
They should describe the general purpose and conditions for each function and its arguments.

In this case for `slotfunc`, your description is true but not informative for users.
Suppose that you are a user of this module. You want to use `LoggingHandler`. Would you think this docstring is helpful? It seems that it can only be used by `LoggerApp`, even if `LoggingHandler` is a public class.

From now on, please write the docstring as general as possible!

My suggestion: "A slot function for the log signal. It will be called with a log message when a log record is emitted."

Aijuh replied:
I agree. Thanks for detailed instruction. My thought was short. I will follow exact format and think in general perspective.

I fixed codes

kangz12345 commented on examples/logger.py:
```diff
@@ -3,20 +3,60 @@
+    Sends a log message to connected function using emit.   
+    """
+
+    def __init__(self, slotfunc):
```
Please annotate type: `slotfunc: Callable[[str], None]`.

kangz12345 replied:
You have to add `Callable` in the `from typing import ...`.

kangz12345 reviewed (changes requested):
Might be the final review. Almost there!

Aijuh committed:
general comment

Aijuh committed:
add Callable

kangz12345 commented on examples/logger.py:
```diff
@@ -28,7 +28,7 @@ class LoggingHandler(logging.Handler):
     """
 
-    def __init__(self, slotfunc):
+    def __init__(self, slotfunc: Callable[[str],None]):
```
Please put a space after a comma!
`Callable[[str], None]`

kangz12345 commented on examples/logger.py:
```diff
@@ -34,7 +34,7 @@ def __init__(self, slotfunc):
 
         Args:
-            slotfunc: A method addLog in class LoggerApp.
+            slotfunc: A slot function which is called when log record emitted.
```
👍

BECATRUE commented on examples/logger.py:
```diff
@@ -85,11 +126,19 @@ def __init__(self, name: str, parent: Optional[QObject] = None):
         self.loggerFrame.clearButton.clicked.connect(self.checkToClear)
         self.confirmFrame = ConfirmClearingFrame()
         self.confirmFrame.confirmed.connect(self.clearLog)
+        self.handler = LoggingHandler(self.addLog)
```
Why did you make `handler` as an attribute of `LoggerApp`?
Is it necessary?

Aijuh replied:
No as far as I check, it doesn't matter whether setting it as local or class variable.

kangz12345 replied:
Maybe setting it as an attribute would be good because it allows users to access the handler object.

BECATRUE commented on examples/logger.py:
```diff
@@ -3,20 +3,60 @@
 from qiwis import BaseApp
 
+
+class Signaller(QObject):
```
I still wonder why this class is necessary?
It has only one signal and I think you can move it into `LoggingHandler`.

BECATRUE replied:
Is there any discussion about this with Jiyong?

Aijuh replied:
Its because pyQt5.Qtcore.pyqtSignal object has no attribute 'connect'. So we need to set Qobject as parent of LoggingHandler to use signal.connect. So we need to set Logging.Handler and Qobject as parent of LoggingHandler, but @kangz12345 said that multiple inheritance may be misleading. 
He pointed out vague meaning for initialization function.

BECATRUE replied:
Oh, I understand now!
Then, how about adding `_` at the begining of the class name? (Just rename `_Signaller`)
It is only for `LoggingHandler`, so it can be treated as a private class.

BECATRUE commented on examples/logger.py:
```diff
@@ -3,20 +3,60 @@
+    def emit(self, record: logging.LogRecord):
+        """Overridden.
+        
+        Emits input signal to connected function.
```
It is a minor suggestion.
How about 'Emits the input signal to the connected function'.

I don't know well which is more appropriate..

BECATRUE reviewed (changes requested):
I left a few reviews!

Aijuh committed:
fix format error

Aijuh commented:
1. Fix error in format
2. In LoggerApp, handler and logger are not class variable.

Aijuh committed:
Update logger.py

Aijuh committed:
test

Aijuh committed:
Update logger.py

Aijuh commented:
Unittest failed with following error.

_E: Failed to fetch http://azure.archive.ubuntu.com/ubuntu/pool/main/libi/libinput/libinput10_1.20.0-1ubuntu0.2_amd64.deb  404  Not Found [IP: 52.252.75.106 80]
E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?___

It seems that fetch link has error. Can I do something to fix it?

Aijuh committed:
Update logger.py

BECATRUE commented:
> Unittest failed with following error.
> 
> _E: Failed to fetch http://azure.archive.ubuntu.com/ubuntu/pool/main/libi/libinput/libinput10_1.20.0-1ubuntu0.2_amd64.deb 404 Not Found [IP: 52.252.75.106 80] E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?___
> 
> It seems that fetch link has error. Can I do something to fix it?

Oh, we also realized this problem in other repos.
Can you add `sudo apt-get update` into both Pylint and unittest in 'Install dependencies'?
The [unittest](https://github.com/snu-quiqcl/iquip/blob/develop/.github/workflows/unittest.yml) in `iquip` will help you!

BECATRUE commented on examples/logger.py:
```diff
@@ -3,20 +3,60 @@
 
 from qiwis import BaseApp
 
+
```
Please remove the empty line.
We usually make only one line after imporing modules.

BECATRUE reviewed (commented):
it looks good!
After resolving the unittest problem, please re-request to Jiyong.

Aijuh committed:
Update logger.py

Aijuh committed:
Fix pylint, unittest

kangz12345 commented on examples/logger.py:
```diff
@@ -3,20 +3,59 @@
+        Connects the slotfunc to the signal.
+
+        Args:
+            slotfunc: A slot function which is called when log record emitted.
```
... when _a_ log record _is_ emitted.

kangz12345 commented on examples/logger.py:
```diff
@@ -3,20 +3,59 @@
+    Sends a log message to connected function using emit.   
+    """
+
+    def __init__(self, slotfunc: Callable[[str], None]):
```
Oh, I'm sorry to change, but `Callable[[str], Any]` would be better since we don't care about the return value of `slotfunc`.

Aijuh committed:
Minor changes

Aijuh commented:
Fixed code with reference to comments.

I will properly use 'a' or 'the'.

kangz12345 commented on examples/logger.py:
```diff
@@ -27,13 +27,13 @@ class LoggingHandler(logging.Handler):
 
         Args:
-            slotfunc: A slot function which is called when log record emitted.
+            slotfunc: A slot function which is called when a log record emitted.
```
Since the logger emits the log records, so "a log record _is_ emitted.", i.e., you should use the passive form using "is".

kangz12345 commented on .github/workflows/unittest.yml:
```diff
@@ -16,6 +16,7 @@ jobs:
         python-version: ${{ matrix.python-version }}
     - name: Install dependencies
       run: |
+        sudo apt-get update
```
This causes a merge conflict. Please change it to `sudo apt-get clean`.
Don't change it for pylint.yml!

Aijuh replied:
I think changing to `sudo apt-get clean` makes unittest error again.

kangz12345 replied:
Hmm, I think so too. 
Since there is no other PR in this repository, let's revert it to `sudo apt-get update` and resolve the merge conflict.
Please update your code, then I will resolve the conflict.

Aijuh replied:
Updated code in unittest.

kangz12345 replied:
Good, I resolved the conflict.

Aijuh committed:
fixed comments

Aijuh commented:
Fixed comments.

Sorry for the repeated grammer error. I will check every comment with the grammer program later.

Fix pylint and unittest.

kangz12345 reviewed (dismissed):
LGTM!

Aijuh committed:
change unittest

kangz12345 committed:
Merge branch 'develop' into Aijuh/181/fix-logger

BECATRUE reviewed (approved):
Great job adjusting to our coding style!

Then, please merge this branch to the develop branch just by clicking the below button😃
